### PR TITLE
refactor(api): update GPIO pins for door/window switch based on CRB revision

### LINKED ISSUE #5289: OT-2 refresh: change door/lid switch sensing based on CRB revision

### PR DESCRIPTION
**Pull request by ahiuchingau**
## overview
As part of the OT-2 refresh project, this PR changes the robot's GPIO configuration based on the central routing board revision detected (see #5477 for the CRB revision PR).

closes #5289

## changelog
- refactor GPIO pin dicts
   - each pin is an object of the GPIOPin class
   - all of the pins are stored in a list, `gpio_list`, which is an instance of the GPIOList class 
- gpios are now configured based on central routing board revision detected
- GPIOCharDev starts monitoring door switch status when the hardware controller is being built
   - the door switch status is monitored by watching the file descriptor associated with the pin

## risk assessment
Low - mainly touches changes GPIOCharDev

### CODE REVIEW

File: api/src/opentrons/drivers/rpi_drivers/gpio.py [hunk 1]
```diff
@@ -25,13 +26,24 @@
...
     'REV_0': 17,
     'REV_1': 27
 }
 
+INPUT_PINS = {
+    BoardRevision.OG: {
+        'BUTTON_INPUT': 5,
+        'WINDOW_DOOR_SW': 20
+        },
+    BoardRevision.A: {
```
Comment (sfoster1) [inline api/src/opentrons/drivers/rpi_drivers/gpio.py:39]: we should probably have defaults for b and c here too just in case

File: api/src/opentrons/drivers/rpi_drivers/gpio.py [hunk 2]
```diff
@@ -25,13 +26,24 @@
...
-    'WINDOW_INPUT': 20,
+REV_PINS = {
     'REV_0': 17,
     'REV_1': 27
 }
 
+INPUT_PINS = {
+    BoardRevision.OG: {
+        'BUTTON_INPUT': 5,
+        'WINDOW_DOOR_SW': 20
```
Comment (sfoster1) [inline api/src/opentrons/drivers/rpi_drivers/gpio.py:37]: i think it's probably best to have definition for all the same pins (so, make this an enum) and either allow them to be none or have, in this case, the window and door switch both be the same gpio.

File: api/src/opentrons/drivers/rpi_drivers/gpio.py [hunk 3]
```diff
@@ -73,17 +85,33 @@ def _retry_request_line(retries: int = 0):
...
+        for name, offset in REV_PINS.items():
             lines[offset] = self._request_line(
                 offset, name, gpiod.LINE_REQ_DIR_IN)
+        return lines
+
+    def config_by_board_rev(self, board_rev: BoardRevision):
+        MODULE_LOG.info(
+            "Configuring GPIOs by central routing roard revision")
+        # gpio config is defaulted to 2.1 if board rev is unknown
+        board_rev = BoardRevision.OG if (
```
Comment (sfoster1) [inline api/src/opentrons/drivers/rpi_drivers/gpio.py:100]: To me, this object should do its best to offer the same api regardless of revision. At some point that may be impossible, but for now it's definitely a thing we can do. That means having the same lines all the time, and it means under-the-hood merging implementations. For instance, having a line for both window and door switch even on the original rev, and just mapping them to the same thing internally. We want this to be the layer where we do that hard work

File: api/src/opentrons/drivers/rpi_drivers/gpio.py [hunk 4]
```diff
@@ -152,10 +185,20 @@ def get_rail_lights(self) -> bool:
 
     def read_button(self) -> bool:
         # button is normal-HIGH, so invert
-        return not bool(self._read(INPUT_PINS['BUTTON_INPUT']))
+        return not bool(self._read(
+            INPUT_PINS[BoardRevision.OG]['BUTTON_INPUT']))
 
     def read_window_switches(self) -> bool:
-        return bool(self._read(INPUT_PINS['WINDOW_INPUT']))
+        return bool(self._read(
```
Comment (sfoster1) [inline api/src/opentrons/drivers/rpi_drivers/gpio.py:192]: for instance, we should always have `read_door_switch` and `read_window_switch`; they're just backed by the same gpio.

File: api/src/opentrons/drivers/rpi_drivers/gpio.py [hunk 5]
```diff
@@ -25,13 +26,24 @@
...
+        'BUTTON_INPUT': 5,
+        'WINDOW_DOOR_SW': 20
+        },
+    BoardRevision.A: {
+        'BUTTON_INPUT': 5,
+        'DOOR_SW_FILT': 12,
+        'WINDOW_SW_FILT': 16,
+        'WINDOW_DOOR_SW': 20
+        }
+}
```
Comment (sfoster1) [inline api/src/opentrons/drivers/rpi_drivers/gpio.py:45]: At this point we should also hoist the gpio line definitions up into a types module so it doesn't have to be defined in two places

File: api/src/opentrons/hardware_control/controller.py [hunk 6]
```diff
@@ -90,6 +91,16 @@ def determine_board_revision(self) -> BoardRevision:
...
 
+    async def monitor_door_switch_state(self, api):
+        while True:
+            await asyncio.sleep(0.5)
+            state = self.gpio_chardev.read_window_switches()
+            if api._door_state.value != state:
+                api._door_state = DoorState.by_bool(state)
+                MODULE_LOG.info(
+                    f'Updating the window switch status: {api._door_state}')
+                # TODO: pause protocol if it's running
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/controller.py:102]: we don't have to have a todo here, that'll be a separate pr

File: api/src/opentrons/hardware_control/controller.py [hunk 7]
```diff
@@ -83,13 +84,22 @@ def determine_board_revision(self) -> BoardRevision:
...
         except Exception:
             MODULE_LOG.exception(
                 'Unexpected error from reading central routing board '
                 'revision bits')
-            return BoardRevision.UNKNOWN
+        return BoardRevision.UNKNOWN
+
+    async def monitor_door_switch_state(self, api):
+        while True:
+            await asyncio.sleep(0.5)
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/controller.py:96]: Rather than polling, it would be really nice to use libgpio's event monitoring here using `LineEvent`.

File: api/src/opentrons/drivers/rpi_drivers/types.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
+import enum
+from collections import defaultdict
+from typing import Dict, List, Union
+from opentrons.hardware_control.types import BoardRevision
+
+
```
Comment (amitlissack) [inline api/src/opentrons/drivers/rpi_drivers/types.py:6]: Are you sure enums are the best way to model pins? Perhaps a Pin class that has a name, a pin number, in/out, and revision mapping would make sense here? There could be a collection that has all the pins and filtering functions to get in/out or by revision etcetera. 

What do you think?

File: api/src/opentrons/drivers/rpi_drivers/types.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+def group_by_gpio(
+        input_dict: Dict[str, int]) -> Dict[int, List[str]]:
+    d: Dict[int, List[str]] = defaultdict(list)
+    for key, value in sorted(input_dict.items()):
+        d[value].append(key)
+    return d
+
+
+GPIOPins = Union[OutputPins, RevPins, InputPins]
```
Comment (amitlissack) [inline api/src/opentrons/drivers/rpi_drivers/types.py:68]: It's typical to use a singular noun when naming an Enum. 

Seeing references to the type `GPIOPIns` I think it is a type that is a collection of Pin-like objects.

File: api/src/opentrons/hardware_control/controller.py [hunk 10]
```diff
@@ -25,6 +25,19 @@
 MODULE_LOG = logging.getLogger(__name__)
 
 
+def event_callback(q: asyncio.Queue):
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/controller.py:28]: this should definitely be an underscore-prefixed name to make it apparent it's internal

File: api/src/opentrons/hardware_control/controller.py [hunk 11]
```diff
@@ -25,6 +25,19 @@
...
+    try:
+        q.put_nowait(QueueEvents.EVENT_RECEIVED)
+    except asyncio.QueueFull:
+        MODULE_LOG.warning(
+            'Event queue is full and can no longer receive new events')
+
+
+class QueueEvents(enum.Enum):
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/controller.py:36]: probably call this GpioQueueEvents

File: api/src/opentrons/hardware_control/types.py [hunk 12]
```diff
@@ -2,9 +2,13 @@
 import asyncio
 import enum
 import logging
-from typing import Tuple
+from typing import Tuple, TYPE_CHECKING
 from opentrons import types as top_types
 
+if TYPE_CHECKING:
+    import gpiod  # type: ignore
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/types.py:9]: this will fail if you're not on linux. this should really be only in the gpio chardev module, and nothing outside of that should have to care about gpiod

File: api/src/opentrons/main.py [hunk 13]
```diff
@@ -72,9 +72,9 @@ async def initialize_robot() -> ThreadManager:
 
     async def blink():
         while True:
-            await hardware.set_lights(button=True)
+            hardware.set_lights(button=True)
```
Comment (sfoster1) [inline api/src/opentrons/main.py:75]: is this on purpose? I think this should indeed be `await` since it's from an async client

File: api/src/opentrons/drivers/rpi_drivers/types.py [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+        ref = {
+            BoardRevision.OG: self.rev_og,
+            BoardRevision.A: self.rev_a,
+            BoardRevision.B: self.rev_b,
+            BoardRevision.C: self.rev_c,
+            BoardRevision.UNKNOWN: self.rev_og}
+        return ref[board_rev]
+
+
+class GPIOList(list):
```
Comment (sfoster1) [inline api/src/opentrons/drivers/rpi_drivers/types.py:54]: The interaction pattern for `GPIOList` is actually really cool but I'm not sure that I agree with making it a list. The interaction pattern doesn't look like a list to me, and we don't really actually use it as a list, other than the initializer, and we can just store a list internally. We only use the accessors you've listed here like `__getattr__`. 

Le'ts just make this not inherit from list and have an explicit initializer that stores the list internally and we're good. Inheriting from core python types is usually not a great idea (and if you really really need to, you should use the `collections` equivalent, which is designed for the use case, eg [UserList](https://docs.python.org/3/library/collections.html#collections.UserList)

File: api/src/opentrons/hardware_control/controller.py [hunk 15]
```diff
@@ -279,3 +264,4 @@ def __del__(self):
             loop = asyncio.get_event_loop()
             if loop.is_running() and self._module_watcher:
                 self._module_watcher.close()
+        self.gpio_chardev.quit_monitoring()
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/controller.py:267]: this should be wrapped in a `if hasattr(self, 'gpio_chardev')` just to prevent the case where an early exception happens before the gpio chardev is initialized and triggers another exception here.

File: api/src/opentrons/drivers/rpi_drivers/gpio_simulator.py [hunk 16]
```diff
@@ -77,30 +73,51 @@ def set_isp_pin(self, on: bool = True):
...
+        if val:
+            return DoorState.CLOSED
         else:
-            raise RevisionPinsError
+            return DoorState.OPEN
 
-    def release_line(self, offset: int):
-        self.lines.pop(offset)
-        self._values.pop(offset)
+    async def monitor_door_switch_state(
```
Comment (sfoster1) [inline api/src/opentrons/drivers/rpi_drivers/gpio_simulator.py:114]: this should probably leave a persistent task running (but we don't need a full queue - we can use a single Event that only controls quit, triggered by `quit_monitoring`)

File: api/src/opentrons/drivers/rpi_drivers/gpio_simulator.py [hunk 17]
```diff
@@ -77,30 +73,51 @@ def set_isp_pin(self, on: bool = True):
...
+        return bool(self._read(gpio_list.door_sw_filt))
 
     def read_revision_bits(self) -> Tuple[bool, bool]:
-        if pathlib.Path(DTOVERLAY_PATH).exists():
-            return (bool(self._read(INPUT_PINS['REV_0'])),
-                    bool(self._read(INPUT_PINS['REV_1'])))
+        return (bool(self._read(gpio_list.rev_0)),
+                bool(self._read(gpio_list.rev_1)))
+
+    def get_door_switches_fd(self) -> int:
```
Comment (sfoster1) [inline api/src/opentrons/drivers/rpi_drivers/gpio_simulator.py:104]: This isn't necessary anymore right?

File: api/tests/opentrons/hardware_control/test_gpio.py [hunk 18]
```diff
@@ -1,9 +1,30 @@
...
+
+async def test_gpio_door_state(loop):
+    # API.door_state would change based on the value read from
+    # the window_door_sw pin
+    hw_api = await hc.API.build_hardware_simulator(
+        loop=loop)
+
+    # high -> door is open
+    hw_api._backend.gpio_chardev.set_low(gpio_list.window_door_sw)
+    await hw_api._backend.gpio_chardev.monitor_door_switch_state(
```
Comment (sfoster1) [inline api/tests/opentrons/hardware_control/test_gpio.py:22]: once we make the simulator monitor persistent we can just check that the callback fires here

**Comment by sfoster1** [comment]
works on my robot!
